token/js/test/e2e-2022/tokenMetadata.test.ts

```diff
@@ -0,0 +1,181 @@
+    emitTokenMetadata,
+} from '../../src';
+import { TEST_PROGRAM_ID, newAccountWithLamports, getConnection } from '../common';
+
+const TEST_TOKEN_DECIMALS = 2;
+
+// TODO: Remove or move to src. Temp function so we can repack to use spl-token-metadata
+const lengthBuffer = (buffer: Buffer | Uint8Array): Buffer => {
+    const length = Buffer.alloc(4);
+    length.writeUIntLE(buffer.length, 0, 4);
+    return length;
+};
+
+describe('Token Metadata', () => {
+    // TODO @mistersimon:- Change to instruction creator once merged
+    // const EXTENSIONS = [ExtensionType.MetadataPointer];
+    // const mintLen = getMintLen(EXTENSIONS);
+    const mintLen = 234;
+
+    let connection: Connection;
+    let payer: Signer;
+    let mint: Keypair;
+    const authority = Keypair.fromSecretKey(
+        new Uint8Array([
+            118, 177, 37, 231, 15, 88, 210, 92, 79, 231, 202, 22, 11, 15, 121, 54, 95, 229, 149, 119, 48, 177, 187, 198,
+            223, 51, 225, 74, 12, 54, 172, 36, 207, 107, 122, 208, 209, 168, 61, 177, 190, 137, 23, 156, 84, 32, 34, 82,
+            158, 176, 55, 51, 236, 66, 130, 167, 118, 31, 120, 107, 100, 192, 147, 10,
+        ])
+    );
+
+    before(async () => {
+        connection = await getConnection();
+        payer = await newAccountWithLamports(connection, 1000000000);
+    });
+
+    beforeEach(async () => {
+        mint = Keypair.generate();
+
+        // TODO @mistersimon:- Change to instruction creator once merged
+        const metadataPointerInstruction = new TransactionInstruction({
+            keys: [{ pubkey: mint.publicKey, isSigner: false, isWritable: true }],
+            programId: TEST_PROGRAM_ID,
+            data: Buffer.concat([new Uint8Array([39, 0]), authority.publicKey.toBuffer(), mint.publicKey.toBuffer()]),
+        });
+
+        const lamports = await connection.getMinimumBalanceForRentExemption(mintLen);
+
+        const transaction = new Transaction().add(
+            SystemProgram.createAccount({
+                fromPubkey: payer.publicKey,
+                newAccountPubkey: mint.publicKey,
+                space: mintLen,
+                lamports: lamports * 10,
+                programId: TEST_PROGRAM_ID,
+            }),
+            metadataPointerInstruction,
+            createInitializeMintInstruction(
+                mint.publicKey,
+                TEST_TOKEN_DECIMALS,
+                authority.publicKey,
+                null,
+                TEST_PROGRAM_ID
+            )
+        );
+
+        await sendAndConfirmTransaction(connection, transaction, [payer, mint], undefined);
+
+        await initializeTokenMetadata(
+            connection,
+            payer,
+            authority.publicKey,
+            mint.publicKey,
+            authority,
+            'name',
+            'symbol',
+            'uri',
+            undefined,
+            undefined,
+            TEST_PROGRAM_ID
+        );
+    });
+
+    const getTokenMetadata = async () => {
```

Comment:
1. Do we need to a function like this in the library? Couldn't see anything similar - Would it be considered an action? 

1. I don't understand why ExtensionType needs to be 19 (which is GroupPointer) - Is it correct, or just a coincidence it works?

1. Seems like we can't use the `unpack` in the `@solana/spl-token-metadata`  library, because its not a tlv. Should we refactor `@solana/spl-token-metadata` to expose a function to decode the pure data? Or create the same functionality in this library?

Reply:
> 1. Do we need to a function like this in the library? Couldn't see anything similar - Would it be considered an action?
Yeah we probably do, even though I don't believe the Token Rust client has this yet!

> 2. I don't understand why ExtensionType needs to be 19 (which is GroupPointer) - Is it correct, or just a coincidence it works?
I count `TokenMetadata` as 19.

> 3. Seems like we can't use the `unpack` in the `@solana/spl-token-metadata`  library, because its not a tlv. Should we refactor `@solana/spl-token-metadata` to expose a function to decode the pure data? Or create the same functionality in this library?
Ahh okay, I think I see the problem here. Token2022 isn't using 8-byte discriminators, so we can't fix the `unpack` function in `@solana/spl-token-metadata` to use 8-byte discriminators. We'll have to modify that function so it can be used here, or refactor the serialization bits of the `TokenMetadata` state in the JS lib to be more friendly with the Rust trait `VariableLenPack`:

https://github.com/solana-labs/solana-program-library/blob/f542e00a8905eaf3f55dea1c08ca4787138e1788/token-metadata/interface/src/state.rs#L99-L109